sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala

```diff
@@ -328,3 +281,222 @@ object FrameBoundaryExtractor {
+    this(input, offset, Literal(null))
+
+  def this(input: Expression) =
+    this(input, Literal(1), Literal(null))
+
+  def this() = this(Literal(null), Literal(1), Literal(null))
+
+  val offsetSign = -1
+}
+
+abstract class AggregateWindowFunction extends DeclarativeAggregate with WindowFunction {
+  self: Product =>
+  override val frame = SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
+  override def dataType: DataType = IntegerType
+  override def nullable: Boolean = false
+  override val mergeExpressions = Nil // TODO how to deal with this?
+}
+
+abstract class RowNumberLike extends AggregateWindowFunction {
+  override def children: Seq[Expression] = Nil
+  override def inputTypes: Seq[AbstractDataType] = Nil
+  protected val zero = Literal(0)
+  protected val one = Literal(1)
+  protected val rowNumber = AttributeReference("rowNumber", IntegerType, false)()
+  override val aggBufferAttributes: Seq[AttributeReference] = rowNumber :: Nil
+  override val initialValues: Seq[Expression] = zero :: Nil
+  override val updateExpressions: Seq[Expression] = Add(rowNumber, one) :: Nil
+}
+
+/**
+ * A [[SizeBasedWindowFunction]] needs the size of the current window for its calculation.
+ */
+trait SizeBasedWindowFunction extends AggregateWindowFunction {
+  /** The partition size attribute, this is here to prevent us from creating an attribute on the
+    * executor side. */
+  val size = AttributeReference("size", IntegerType, false)()
+
+  /** Set the window size expression. */
+  def withSize(n: Expression): SizeBasedWindowFunction
+}
+
+case class RowNumber() extends RowNumberLike {
+  override val evaluateExpression = Cast(rowNumber, IntegerType)
+}
+
+case class CumeDist(n: Expression)
+    extends RowNumberLike with SizeBasedWindowFunction {
+  def this() = this(Literal(0))
+  override def dataType: DataType = DoubleType
+  override def withSize(n: Expression): CumeDist = CumeDist(n)
+  override val frame = SpecifiedWindowFrame(RangeFrame, UnboundedPreceding, CurrentRow)
+  override val evaluateExpression = Divide(Cast(rowNumber, DoubleType), Cast(n, DoubleType))
+}
+
+case class NTile(buckets: Expression, n: Expression)
+    extends RowNumberLike with SizeBasedWindowFunction {
+  def this() = this(Literal(1), Literal(0))
+  def this(buckets: Expression) = this(buckets, Literal(0))
+
+  // Validate buckets.
+  buckets.eval() match {
+    case b: Int if b > 0 => // Ok
+    case x => throw new AnalysisException(
+      "Buckets expression must be a foldable positive integer expression: $x")
+  }
+
+  override def withSize(n: Expression): NTile = NTile(buckets, n)
+  private val bucket = AttributeReference("bucket", IntegerType, false)()
+  private val bucketThreshold = AttributeReference("bucketThreshold", IntegerType, false)()
+  private val bucketSize = AttributeReference("bucketSize", IntegerType, false)()
+  private val bucketsWithPadding = AttributeReference("bucketsWithPadding", IntegerType, false)()
```

Comment:
They are not, they are only constant (foldable if you will) for a single group. They are both dependent on size of the current group, the ```n``` expression. The value of the ```n``` expression (and the dependent ```bucketSize``` and ```bucketsWithPadding```) will be updated when we start to process the new group.